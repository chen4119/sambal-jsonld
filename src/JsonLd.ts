import {JSONLD_ID, JSONLD_CONTEXT, JSONLD_TYPE, JSONLD_GRAPH, JSONLD_VALUE, JSONLD_BASE} from "./Constants";
import {isObjectLiteral, isUrl, makeAbsoluteIRI, isBlankNodeIRI} from "./Utils";

export type Node = {
    links: Link[],
    data: any,
    isArray: boolean,
    id: string
};

export type Link = {
    edge: string,
    node: Node
}

export type Term = {
    id?: string,
    type?: string
}

class JsonLd {
    // private vocab: string;
    private blankNodeIndex = 1;
    private contextRef: string;
    private base: string;
    private contextTermMap: Map<string, Term> = new Map<string, Term>();
    private graph: Map<string, Node> = new Map<string, Node>();
    constructor(private json: object, private context?: any) {
        this.parse(json, context);
    }

    async resolveJson(fetcher: (url) => Promise<any>) {
        if (this.isReferencingOtherNode(this.json)) {
            const link = this.json[JSONLD_ID];
            return await this.fetchNode(link, fetcher);
        }
        for (const propName of this.contextTermMap.keys()) {
            const term: Term = this.contextTermMap.get(propName);
            if (term.type === JSONLD_ID) {
                const link = this.json[propName];
                const propValue = await this.fetchNode(link, fetcher);
                this.json[propName] = propValue;
            }
        }
        return this.json;
    }

    get baseIRI(): string {
        return this.base;
    }

    get graphMap(): Map<string, Node> {
        return this.graph;
    }

    flatten() {
        const flattenNodeMap: Map<string, object> = new Map<string, object>();
        for (const nodeId of this.graph.keys()) {
            const nodeObj = this.getCondensedNode(nodeId);
            flattenNodeMap.set(nodeId, nodeObj);
        }
        return flattenNodeMap;
    }

    getCondensedNode(nodeId: string) {
        const node: Node = this.graph.get(nodeId);
        const obj = {};
        for (const link of node.links) {
            const propName = this.getCondensedPropName(link.edge);
            this.setNodeObjectProp(obj, propName, link.node);
        }
        // set generated id first.  If an actual id is specified, will override from node.data
        if (node.id) {
            obj[JSONLD_ID] = node.id;
        }
        if (node.data) {
            for (const propName of Object.keys(node.data)) {
                obj[propName] = node.data[propName];
            }
        }
        return obj;
    }

    private setNodeObjectProp(obj: object, propName: string, node: Node) {
        const propValue = node.id ? {[JSONLD_ID]: node.id} : node.data;
        if (node.isArray) {
            obj[propName] = Array.isArray(obj[propName]) ? [...obj[propName], propValue] : [propValue];
        } else {
            obj[propName] = propValue;
        }
    }

    private getCondensedPropName(absoluteIRI: string) {
        if (isUrl(absoluteIRI)) {
            for (const termKey of this.contextTermMap.keys()) {
                const term = this.contextTermMap.get(termKey);
                if (absoluteIRI === term.id) {
                    return termKey;
                } else if (absoluteIRI.startsWith(term.id)) {
                    return `${termKey}:${this.getIRISuffix(term.id, absoluteIRI)}`;
                }
            }
            if (this.contextRef && absoluteIRI.startsWith(this.contextRef)) {
                return this.getIRISuffix(this.contextRef, absoluteIRI);
            }
        }
        return absoluteIRI;
    }

    private getIRISuffix(prefix: string, absoluteIRI: string) {
        const suffix = absoluteIRI.substring(prefix.length);
        if (suffix.startsWith("/")) {
            return suffix.substring(1);
        }
        return suffix;
    }

    private async fetchNode(link: string, fetcher: (url) => Promise<any>) {
        if (this.graph.has(link)) {
            return this.graph.get(link);
        }
        let url = link;
        if (isBlankNodeIRI(link)) {
            url = link.substring(2);
        }
        const externalObj = await fetcher(url);
        if (Array.isArray(externalObj)) {
            for (const obj of externalObj) {
                const found = this.findNodeById(link, obj);
                if (found) {
                    return found;
                }
            }
        } else if (isObjectLiteral(externalObj)) {
            const found = this.findNodeById(link, externalObj);
            if (found) {
                return found;
            }
        }
        return externalObj;
    }

    private findNodeById(id: string, json: object) {
        if (json[JSONLD_GRAPH]) {
            const jsonld = new JsonLd(json);
            if (jsonld.graphMap.has(id)) {
                return jsonld.getCondensedNode(id);
            }
        }
        return null;
    }

    private parse(json: object, context?: any) {
        if (json[JSONLD_CONTEXT]) {
            this.parseContext(json[JSONLD_CONTEXT]);
        } else if (context) {
            this.parseContext(context);
        }
        if (json[JSONLD_GRAPH]) {
            this.parseGraph(json[JSONLD_GRAPH]);
        } else {
            // since no graph, use empty graph map
            const emptyGraphMap = new Map<string, any>();
            if (Array.isArray(json)) {
                for (const node of json) {
                    this.parseGraphNode(emptyGraphMap, node);
                }
            } else {
                this.parseGraphNode(emptyGraphMap, json);
            }
        }
    }

    private parseContext(context: any) {
        if (typeof(context) === "string") {
            this.contextRef = context;
        } else if (Array.isArray(context)) {
            context.forEach(c => this.parseContext(c));
        } else if (isObjectLiteral(context)) {
            this.parseContextObject(context);
        }
    }

    private parseContextObject(context: object) {
        for (const key of Object.keys(context)) {
            const contextValue = context[key];
            if (key === JSONLD_BASE) {
                this.base = contextValue;
            } else if (typeof(contextValue) === "string") {
                this.contextTermMap.set(key, {id: contextValue});
            } else { // context value is an object literal
                if (contextValue[JSONLD_ID]) {
                    this.contextTermMap.set(key, {
                        id: contextValue[JSONLD_ID],
                        type: contextValue[JSONLD_TYPE]
                    });
                }
            }
        }
    }

    private parseGraph(graph: any[]) {
        const graphMap: Map<string, any> = new Map<string, any>();
        for (const data of graph) {
            this.recursivelyVisitAllNodes(graphMap, data);
        }
        for (const data of graph) {
            this.parseGraphNode(graphMap, data);
        }
    }

    private recursivelyVisitAllNodes(graphMap: Map<string, any>, data: any) {
        if (this.isReferencingOtherNode(data)) {
            return;
        }
        if (data[JSONLD_ID]) {
            graphMap.set(data[JSONLD_ID], data);
        }
        for (const key of Object.keys(data)) {
            const value = data[key];
            if (Array.isArray(value)) {
                value.forEach(child => this.recursivelyVisitAllNodes(graphMap, child));
            } else if (isObjectLiteral(value)) {
                this.recursivelyVisitAllNodes(graphMap, value);
            }
        }
    }

    private parseGraphNode(graphMap: Map<string, any>, data: any): Node {
        const nodeId = data[JSONLD_ID] ? data[JSONLD_ID] : this.getNextBlankNodeId();
        const isReferencingOtherNode = this.isReferencingOtherNode(data);
        if (this.graph.has(nodeId)) {
            return this.graph.get(nodeId);
        } else if (isReferencingOtherNode && graphMap.has(nodeId)) {
            return this.parseGraphNode(graphMap, graphMap.get(nodeId));
        }

        const node: Node = {links: [], data: {}, isArray: false, id: nodeId};
        if (!isReferencingOtherNode) {
            this.graph.set(nodeId, node);
        }
        for (const propName of Object.keys(data)) {
            const absoluteIRI = this.getAbsoluteIRI(propName);
            if (isUrl(absoluteIRI)) {
                this.addNodeLink(graphMap, node, absoluteIRI, data[propName], false);
            } else {
                node.data[propName] = data[propName];
            }
        }
        return node;
    }

    private addNodeLink(graphMap: Map<string, any>, node: Node, absoluteIRI: string, value: any, isArray: boolean) {
        if (Array.isArray(value)) {
            value.forEach(child => this.addNodeLink(graphMap, node, absoluteIRI, child, true));
            return;
        }
        let targetNode;
        if (isObjectLiteral(value)) {
            targetNode = this.parseGraphNode(graphMap, value);
            targetNode.isArray = isArray;
        } else {
            targetNode = {links: [], data: value, isArray: isArray};
        }
        node.links.push({
            edge: absoluteIRI,
            node: targetNode
        });
    }

    private getNextBlankNodeId() {
        return `_:${this.blankNodeIndex++}`;
    }

    private isReferencingOtherNode(data: object) {
        return Object.keys(data).length === 1 && data[JSONLD_ID];
    }

    private getAbsoluteIRI(propKey: string) {
        // json-ld keyword
        if (propKey.startsWith("@")) {
            return propKey;
        }

        if (this.contextTermMap.has(propKey)) {
            return this.contextTermMap.get(propKey).id;
        } else if (isUrl(propKey)) {
            return propKey; // already absolute iri
        } else if (propKey.indexOf(":") >= 0) {
            const splitted = propKey.split(":");
            const prefix = splitted[0];
            const suffix = splitted[1];
            const context = this.contextTermMap.get(prefix);
            if (context.id) {
                return makeAbsoluteIRI(context.id, suffix);
            }
        } else if (this.contextRef) {
            return makeAbsoluteIRI(this.contextRef, propKey);
        }
        return propKey;
    }
}

export default JsonLd;