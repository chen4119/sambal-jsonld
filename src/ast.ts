import ts from "typescript";

export const EXPORT_MODIFIER = ts.createModifier(ts.SyntaxKind.ExportKeyword);

export function makeValueLiteral(value: any): ts.Expression {
    if (Array.isArray(value)) {
        return arrayToArrayLiteral(value);
    }
    switch (typeof(value)) {
        case "string":
            return makeStringLiteral(value);
        case "number":
            return makeNumericLiteral(value);
        case "boolean":
            return makeBooleanLiteral(value);
        case "object":
            return objectToObjectLiteral(value);
        case "undefined":
            return makeIdentifier("undefined");    
        default:
            throw new Error(`Unrecognized type: ${typeof(value)}`);
    }
}

export function objectToObjectLiteral(object: any): ts.ObjectLiteralExpression {
    const properties = [];
    for (const prop of Object.keys(object)) {
        properties.push(makePropertyAssignment(prop, makeValueLiteral(object[prop])));
    }
    return makeObjectLiteral(properties);
}

export function arrayToArrayLiteral(array: any[]): ts.ArrayLiteralExpression {
    const values = array.map((v) => makeValueLiteral(v));
    return makeArrayLiteral(values);
}

export function makeObjectLiteral(properties): ts.ObjectLiteralExpression {
    return ts.createObjectLiteral(properties, true);
}

export function makeArrayLiteral(values: ts.Expression[]): ts.ArrayLiteralExpression {
    return ts.createArrayLiteral(values, true);
}

export function makePropertyAssignment(propName: string, initializer): ts.PropertyAssignment{
    return ts.createPropertyAssignment(
        makeIdentifier(propName),
        initializer
    );
}

export function makeIdentifier(text: string): ts.Identifier {
    return ts.createIdentifier(text);
}

export function makeStringLiteral(text: string): ts.StringLiteral {
    return ts.createStringLiteral(text);
}

export function makeNumericLiteral(number: number): ts.NumericLiteral {
    return ts.createNumericLiteral(String(number));
}

export function makeBooleanLiteral(bool: boolean): ts.BooleanLiteral {
    return bool ? ts.createTrue() : ts.createFalse();
}

export function makeVariableStatement(modifiers: ts.Modifier[], variableName: string, variableInitializer, flag = ts.NodeFlags.Const): ts.VariableStatement {
    return ts.createVariableStatement(
        modifiers,
        ts.createVariableDeclarationList(
            [ts.createVariableDeclaration(
                makeIdentifier(variableName),
                undefined,
                variableInitializer
            )], flag
        )
    );
}

export function makeEnum(modifiers: ts.Modifier[], enumName: string, values: string[]) {
    return ts.createEnumDeclaration(
        undefined,
        modifiers,
        makeIdentifier(enumName),
        values.map((v) => ts.createEnumMember(makeIdentifier(v), undefined))
    );
}

export function makeNew(expression: ts.Expression, args: ts.Expression[]): ts.NewExpression {
    return ts.createNew(
        expression,
        undefined,
        args
    );
}
