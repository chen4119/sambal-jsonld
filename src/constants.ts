export const JSONLD_CONTEXT = "@context";
export const JSONLD_TYPE = "@type";
export const JSONLD_ID = "@id";
export const JSONLD_GRAPH = "@graph";
export const JSONLD_VALUE = "@value";
export const JSONLD_BASE = "@base";

export const SCHEMA_CONTEXT = "https://schema.org";

export const SAMBAL_NAME = "_name";
export const SAMBAL_PARENT = "_parent";
export const SAMBAL_VALUES = "_values";

// Schema.org datatypes
export const SCHEMA_TEXT = `${SCHEMA_CONTEXT}/Text`;
export const SCHEMA_CSS_SELECTOR = `${SCHEMA_CONTEXT}/CssSelectorType`;
export const SCHEMA_XPATH = `${SCHEMA_CONTEXT}/XPathType`;

export const SCHEMA_NUMBER = `${SCHEMA_CONTEXT}/Number`;
export const SCHEMA_INTEGER = `${SCHEMA_CONTEXT}/Integer`;
export const SCHEMA_FLOAT = `${SCHEMA_CONTEXT}/Float`;

export const SCHEMA_DATE = `${SCHEMA_CONTEXT}/Date`;
export const SCHEMA_DATETIME = `${SCHEMA_CONTEXT}/DateTime`;
export const SCHEMA_TIME = `${SCHEMA_CONTEXT}/Time`;
export const SCHEMA_BOOL = `${SCHEMA_CONTEXT}/Boolean`;
export const SCHEMA_URL = `${SCHEMA_CONTEXT}/URL`;

export const SCHEMA_PRIMITIVE_SET = new Set([
    SCHEMA_TEXT,
    SCHEMA_CSS_SELECTOR,
    SCHEMA_XPATH,
    SCHEMA_NUMBER,
    SCHEMA_INTEGER,
    SCHEMA_FLOAT,
    SCHEMA_DATE,
    SCHEMA_DATETIME,
    SCHEMA_TIME,
    SCHEMA_BOOL,
    SCHEMA_URL
]);

export const SCHEMA_ENUMERATION = `${SCHEMA_CONTEXT}/Enumeration`;
