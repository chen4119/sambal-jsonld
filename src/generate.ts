import SchemaGenerator from "./SchemaGenerator";


const generator = new SchemaGenerator();
generator.writeSchema("./all-layers.jsonld", "./src/schema.ts");