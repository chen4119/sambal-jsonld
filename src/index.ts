
export {hydrateJsonLd} from "./operators/hydrateJsonLd";
export {toSchemaOrgJsonLd} from "./operators/toSchemaOrgJsonLd";
export {toJsonLdGraph} from "./operators/toJsonLdGraph";
export {graphToCondensedJsonLd} from "./operators/graphToCondensedJsonLd";
export {getSchemaOrgType, isSchemaOrgType} from "./Utils";
export {
    JSONLD_ID,
    JSONLD_TYPE,
    JSONLD_CONTEXT,
    JSONLD_GRAPH,
    JSONLD_VALUE,
    SCHEMA_CONTEXT,
    SCHEMA_TEXT,
    SCHEMA_CSS_SELECTOR,
    SCHEMA_XPATH,
    SCHEMA_NUMBER,
    SCHEMA_INTEGER,
    SCHEMA_FLOAT,
    SCHEMA_DATE,
    SCHEMA_DATETIME,
    SCHEMA_TIME,
    SCHEMA_BOOL,
    SCHEMA_URL,
    SAMBAL_ID,
    SAMBAL_NAME,
    SAMBAL_PARENT,
    SAMBAL_VALUES
} from "./Constants";