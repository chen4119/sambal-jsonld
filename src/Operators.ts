import moment from 'moment';
import {schemaMap} from "./Schema";
import JsonLd, {Node} from "./JsonLd";
import {
    SAMBAL_PARENT,
    SCHEMA_CONTEXT,
    JSONLD_CONTEXT,
    JSONLD_TYPE,
    JSONLD_GRAPH,
    JSONLD_ID,
    SAMBAL_NAME,
    SAMBAL_ID,
    SCHEMA_PRIMITIVE_SET,
    SCHEMA_ENUMERATION,
    SCHEMA_NUMBER,
    SCHEMA_INTEGER,
    SCHEMA_BOOL,
    SCHEMA_TEXT,
    SCHEMA_URL,
    SCHEMA_DATETIME,
    SCHEMA_DATE,
    SCHEMA_TIME
} from "./Constants";
import {isObjectLiteral, isUrl} from "./Utils";

const DATE_FORMAT_REGEX = /^\d{4}-\d{1,2}-\d{1,2}$/;
const TIME_FORMAT_REGEX = /^\d{1,2}:\d{1,2}(:\d{1,2})?$/;

export function isSchemaOrgType(typeId: string) {
    return schemaMap.has(typeId.toLowerCase());
}

export function getSchemaOrgType(typeId: string) {
    return schemaMap.get(typeId.toLowerCase());
}

export function getSchemaOrgParentTypes(typeId: string) {
    const parents = [];
    if (isSchemaOrgType(typeId)) {
        getParentTypes(getSchemaOrgType(typeId), parents);
    }
    return parents;
}

export async function hydrateJsonLd(json: object, fetcher: (url) => Promise<any>, context?: any) {
    if (Array.isArray(json)) {
        const resolvedArray = [];
        for (const item of json) {
            resolvedArray.push(await hydrateJsonLd(item, fetcher, context));
        }
        return resolvedArray;
    } else if (isObjectLiteral(json)) {
        for (const propName of Object.keys(json)) {
            const propValue = await hydrateJsonLd(json[propName], fetcher, context);
            json[propName] = propValue;
        }
        const jsonld = new JsonLd(json, context);
        return await jsonld.resolveJson(fetcher);
    }
    return json;
}

export function getSchemaOrgGraph(schemaOrgJsonLds: object[]) {
    if (schemaOrgJsonLds.length === 0) {
        return null;
    }
    const jsonLd = new JsonLd(schemaOrgJsonLds, SCHEMA_CONTEXT);
    const graph: Map<string, object> = jsonLd.flatten();
    return {
        [JSONLD_CONTEXT]: SCHEMA_CONTEXT,
        [JSONLD_GRAPH]: [...graph.values()]
    };
}

export function toSchemaOrgJsonLd(json: object, type: string, context?: any) {
    const typeId = `${SCHEMA_CONTEXT}/${type}`;
    if (!isSchemaOrgType(typeId)) {
        throw new Error(`schema.org ${type} not found`);
    }
    const jsonld = new JsonLd(json, context ? [SCHEMA_CONTEXT, context] : SCHEMA_CONTEXT);
    let nodeId;
    if (json[JSONLD_ID]) {
        nodeId = json[JSONLD_ID];
    } else {
        nodeId = '_:1';
    }
    if (jsonld.graphMap.has(nodeId)) {
        const node = jsonld.graphMap.get(nodeId);
        return {
            [JSONLD_CONTEXT]: SCHEMA_CONTEXT,
            ...nodeToSchemaOrgType(node, type)
        };
    }
    return null;
}

function getSchemaOrgProps(schema, typeProps) {
    for (const propName of Object.keys(schema)) {
        if (propName.startsWith("_")) {
            continue;
        }
        const propId = `${SCHEMA_CONTEXT}/${propName}`;
        typeProps[propId] = {
            name: propName,
            types: schema[propName]
        };
    }
}

function nodeToSchemaOrgType(node: Node, type: string) {
    const typeId = `${SCHEMA_CONTEXT}/${type}`;
    const typeSchema = getSchemaOrgType(typeId);
    const typeName = typeSchema[SAMBAL_NAME];
    const parents = [];
    getParentTypes(typeSchema, parents);
    const typeProps = {};
    getSchemaOrgProps(typeSchema, typeProps);
    for (const parentSchema of parents) {
        getSchemaOrgProps(parentSchema, typeProps);
    }
    return populateSchemaOrgType(node, typeName, typeProps);
}

function populateSchemaOrgType(node: Node, typeName: string, typeProps: object) {
    const schemaOrgJsonLd = {
        [JSONLD_TYPE]: typeName
    };

    for (const link of node.links) {
        const prop = typeProps[link.edge];
        if (prop) {
            const hasChildNode = link.node.links.length > 0;
            let propValue = null;
            if (hasChildNode) {
                const classTypeName = getClassTypeName(link.node.data, prop.types);
                // possible not to have a class type
                if (classTypeName) {
                    propValue = nodeToSchemaOrgType(link.node, classTypeName);
                }
            } else if (typeof(link.node.data) !== "undefined" && link.node.data !== null) {
                propValue = validatePrimitiveType(link.node.data, prop.types);
            }
            setSchemaOrgPropValue(schemaOrgJsonLd, prop.name, propValue, link.node.isArray);
        }
    }
    return schemaOrgJsonLd;
}

function setSchemaOrgPropValue(schemaOrgJsonLd: object, propName: string, propValue: any, isArray: boolean) {
    if (propValue) {
        if (isArray) {
            schemaOrgJsonLd[propName] = Array.isArray(schemaOrgJsonLd[propName]) ? [...schemaOrgJsonLd[propName], propValue] : [propValue];
        } else {
            schemaOrgJsonLd[propName] = propValue;
        }
    }
}

/*
function populateSchemaProps(schemaOrgJsonLd: object, jsonld: JsonLd, typeProps: object) {
    for (const propName of Object.keys(jsonld.data)) {
        if (!propName.startsWith("@")) {
            const value = getPropValue(propName, jsonld, typeProps);
            if (value) {
                schemaOrgJsonLd[propName] = value;
            }
        }
    }
}


function getPropValue(propName: string, jsonld: JsonLd, typeProps: object) {
    const termId = jsonld.getTermId(propName);
    const prop = typeProps[termId];
    if (prop) {
        const termValue = jsonld.getTermValue(termId);
        if (typeof(termValue) !== "undefined" && termValue !== null) {
            return validatePropValue(termValue, prop.types);
        }
    }
    return null;
}*/

/*
function validatePropValue(value: any, propTypes: string[]) {
    if (Array.isArray(value)) {
        return value.map(item => validatePropValue(item, propTypes));
    } else if (isObjectLiteral(value)) {
        const classTypeName = getClassTypeName(value, propTypes);
        // possible not to have a class type
        if (classTypeName) {
            return toSchemOrgJsonLd(value, classTypeName, SCHEMA_CONTEXT);
        }
    } else if (value) {
        return validatePrimitiveType(value, propTypes);
    }
    return null;
}*/

function validatePrimitiveType(value: any,  propTypes: string[]) {
    const valueType = getPrimitiveType(value);
    if (!valueType) {
        console.error(`Unknown primitive type for value: ${value}`);
        return null;
    }
    switch (valueType) {
        case SCHEMA_DATETIME:
            return validateDate(value, propTypes);
        case SCHEMA_NUMBER:
            return validateNumber(value, propTypes);
        case SCHEMA_TEXT:
            return validateString(value, propTypes);
        default:
            return propTypes.indexOf(valueType) >= 0 ? value : null;
    }
}

function validateDate(value: Date, propTypes: string[]) {
    const isAllowDateTime = propTypes.indexOf(SCHEMA_DATETIME) >= 0;
    const isAllowDate = propTypes.indexOf(SCHEMA_DATE) >= 0;
    const isAllowTime = propTypes.indexOf(SCHEMA_TIME) >= 0;
    if (isAllowDateTime) {
        return moment(value).toISOString();
    } else if (isAllowDate) {
        return moment(value).format("YYYY-MM-DD");
    } else if (isAllowTime) {
        return moment(value).format("HH:mm:ss");
    }
    return null;
}

function validateNumber(value: number, propTypes: string[]) {
    const isAllowNumber = propTypes.indexOf(SCHEMA_NUMBER) >= 0;
    const isAllowInteger = propTypes.indexOf(SCHEMA_INTEGER) >= 0;
    if (isAllowNumber || (isAllowInteger && Number.isInteger(value))) {
        return value;
    }
    return null;
}

function validateString(value: string, propTypes: string[]) {
    const isAllowDateTime = propTypes.indexOf(SCHEMA_DATETIME) >= 0;
    const isAllowDate = propTypes.indexOf(SCHEMA_DATE) >= 0;
    const isAllowTime = propTypes.indexOf(SCHEMA_TIME) >= 0;
    if (isAllowDateTime || isAllowDate || isAllowTime) {
        const d = moment(value, moment.ISO_8601);
        if (!d.isValid()) {
            return null;
        }
        if ((isAllowDate && DATE_FORMAT_REGEX.test(value)) || (isAllowTime && TIME_FORMAT_REGEX.test(value))) {
            return value;
        }  
        return validateDate(d.toDate(), propTypes);
    }
    return value;
}

function getPrimitiveType(value: any) {
    const typeOf = typeof(value);
    if (moment.isDate(value)) {
        return SCHEMA_DATETIME;
    } else if (typeOf === "number") {
        return SCHEMA_NUMBER;
    } else if (typeOf === "boolean") {
        return SCHEMA_BOOL;
    } else if (typeOf === "string") {
        if (isUrl(value)) {
            return SCHEMA_URL;
        }
        return SCHEMA_TEXT;
    }
    return null; // should not happen
}

// potential has more than one class type, just get first one
function getClassTypeName(json: any, propTypes: string[]) {
    if (json[JSONLD_TYPE]) {
        const typeId = `${SCHEMA_CONTEXT}/${json[JSONLD_TYPE]}`;
        for (const validTypeId of propTypes) {
            const isDescendant = isDescendantOf(typeId, validTypeId);
            if (isDescendant) {
                return json[JSONLD_TYPE];
            }
        }
        return null;
    }
    for (const typeId of propTypes) {
        if (!SCHEMA_PRIMITIVE_SET.has(typeId) && !isEnumeration(typeId)) {
            return getSchemaOrgType(typeId)[SAMBAL_NAME];
        }
    }
    return null;
}

function isDescendantOf(schemaTypeId: string, parentTypeId: string) {
    if (isSchemaOrgType(schemaTypeId) && isSchemaOrgType(parentTypeId)) {
        const schema = getSchemaOrgType(schemaTypeId);
        const parents = [];
        getParentTypes(schema, parents);
        for (const parentSchema of parents) {
            if (parentSchema[SAMBAL_ID] === parentTypeId) {
                return true;
            }
        }
    }
    return false;
}

function isEnumeration(typeId: string) {
    const parentSchemas = getSchemaOrgParentTypes(typeId);
    for (const schema of parentSchemas) {
        if (schema[SAMBAL_ID] === SCHEMA_ENUMERATION) {
            return true;
        }
    }
    return false;
}

function getParentTypes(schema, parents) {
    const schemaParents = schema[SAMBAL_PARENT];
    if (schemaParents) {
        for (const parentId of schemaParents) {
            const parentSchema = getSchemaOrgType(parentId);
            parents.push(parentSchema);
            getParentTypes(parentSchema, parents);
        }
    }
}
