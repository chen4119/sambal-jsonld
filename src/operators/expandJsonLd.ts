import JsonLd from "../JsonLd";
import {
    JSONLD_ID
} from "../constants";

export function expandJsonLd(jsonld: unknown) {
    const jsonLd = new JsonLd(jsonld);
    return jsonLd.getNode(jsonld[JSONLD_ID], true, true);
}