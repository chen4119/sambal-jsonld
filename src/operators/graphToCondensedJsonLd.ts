import JsonLd from "../JsonLd";
import {
    JSONLD_GRAPH
} from "../constants";
import {isObjectLiteral} from "../utils";

export function graphToCondensedJsonLd(jsonldGraph: unknown, id?: string) {
    if (isObjectLiteral(jsonldGraph) && jsonldGraph[JSONLD_GRAPH]) {
        const jsonld = new JsonLd(jsonldGraph);
        if (id) {
            return jsonld.getNode(id, true);
        }
        return jsonld.getGraphRootNodes();
    }
    return null;
}