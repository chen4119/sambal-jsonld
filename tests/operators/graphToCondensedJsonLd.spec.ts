import {graphToCondensedJsonLd} from "../../src/operators/graphToCondensedJsonLd";
import {SCHEMA_CONTEXT, JSONLD_ID} from "../../src/constants";

describe('graphToCondensedJsonLd', () => {
    
    const graph = {
        "@context": "http://schema.org",
        "@graph": [
            {
            "@type": "Person",
            "familyName": "Chen",
            "gender": "Male",
            "givenName": "Wan Chun",
            "name": "Wan Chun Chen",
            "sameAs": [
                "https://github.com/chen4119",
                "https://www.linkedin.com/in/wan-chun-chen-9a95a010"
            ],
            "url": "https://chen4119.me/about.html",
            "@id": "https://chen4119.me/about.html#about"
            },
            {
                "@id": "http://chen4119.me/post2",
                "@type": "BlogPosting",
                headline: "Second post",
                description: "Some description",
                author: {
                    "@id": "http://www.janedoe.com/about",
                    "@type": "Person",
                    "name": "Jane Doe",
                    "telephone": "(425) 123-4567"
                },
                dateCreated: "2019-09-06",
                keywords: ["schema.org"]
            }
        ]
    };

    it("has 2 jsonld", () => {
        const json = graphToCondensedJsonLd(graph);
        expect(json).toMatchSnapshot();
    });

    it("gets Jane Doe jsonld", () => {
        const json = graphToCondensedJsonLd(graph, "http://www.janedoe.com/about");
        expect(json).toMatchSnapshot();
    });

    it("return null", () => {
        const json = graphToCondensedJsonLd(graph, "abc");
        expect(json).toBeNull();
    });

});
