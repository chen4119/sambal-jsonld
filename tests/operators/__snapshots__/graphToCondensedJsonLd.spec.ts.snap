// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphToCondensedJsonLd gets Jane Doe jsonld 1`] = `
Object {
  "@id": "http://www.janedoe.com/about",
  "@type": "Person",
  "name": "Jane Doe",
  "telephone": "(425) 123-4567",
}
`;

exports[`graphToCondensedJsonLd has 2 jsonld 1`] = `
Array [
  Object {
    "@id": "https://chen4119.me/about.html#about",
    "@type": "Person",
    "familyName": "Chen",
    "gender": "Male",
    "givenName": "Wan Chun",
    "name": "Wan Chun Chen",
    "sameAs": Array [
      "https://github.com/chen4119",
      "https://www.linkedin.com/in/wan-chun-chen-9a95a010",
    ],
    "url": "https://chen4119.me/about.html",
  },
  Object {
    "@id": "http://chen4119.me/post2",
    "@type": "BlogPosting",
    "author": Object {
      "@id": "http://www.janedoe.com/about",
      "@type": "Person",
      "name": "Jane Doe",
      "telephone": "(425) 123-4567",
    },
    "dateCreated": "2019-09-06",
    "description": "Some description",
    "headline": "Second post",
    "keywords": Array [
      "schema.org",
    ],
  },
]
`;
